File: scripts/flatten_patch.py
#!/usr/bin/env python3
"""From a drawn loop to a patch — port of scripts/flatten_patch.mjs.

Sampling the loop onto the mesh, flood-filling its inside, chord constraints,
cutting an outline into two panels along a seam."""

from __future__ import annotations

import math

from flatten_mesh import edge_face_map



DEFAULT_LOOP_SPACING = 0.002


# ------------------------------------------------------------------ mesh build


def _barycentric(P, F, face: int, q):
    i, j, k = F[face * 3] * 3, F[face * 3 + 1] * 3, F[face * 3 + 2] * 3
    v0 = (P[j] - P[i], P[j + 1] - P[i + 1], P[j + 2] - P[i + 2])
    v1 = (P[k] - P[i], P[k + 1] - P[i + 1], P[k + 2] - P[i + 2])
    v2 = (q[0] - P[i], q[1] - P[i + 1], q[2] - P[i + 2])
    d00 = v0[0] * v0[0] + v0[1] * v0[1] + v0[2] * v0[2]
    d01 = v0[0] * v1[0] + v0[1] * v1[1] + v0[2] * v1[2]
    d11 = v1[0] * v1[0] + v1[1] * v1[1] + v1[2] * v1[2]
    d20 = v2[0] * v0[0] + v2[1] * v0[1] + v2[2] * v0[2]
    d21 = v2[0] * v1[0] + v2[1] * v1[1] + v2[2] * v1[2]
    den = d00 * d11 - d01 * d01
    b1 = (d11 * d20 - d01 * d21) / den
    b2 = (d00 * d21 - d01 * d20) / den
    return (1.0 - b1 - b2, b1, b2)


def point_key(p) -> str:
    """Identity of a sample for matching across pieces (same floats -> same key)."""
    return f"{math.floor(p[0] * 1e9 + 0.5)},{math.floor(p[1] * 1e9 + 0.5)},{math.floor(p[2] * 1e9 + 0.5)}"


def _lex_less(A, B) -> bool:
    if A[0] != B[0]:
        return A[0] < B[0]
    if A[1] != B[1]:
        return A[1] < B[1]
    return A[2] < B[2]


def extract_patch(mesh, closest, loop_points, seed, spacing: float = DEFAULT_LOOP_SPACING) -> dict:
    """Closed loop on the skin -> faces inside it (flood fill bounded by the faces
    the resampled loop lands on, plus those barrier faces themselves). Segments
    are resampled in a canonical direction so a run two loops share yields
    bit-identical samples in both."""
    P, F = mesh["positions"], mesh["faces"]
    nf = len(F) // 3
    samples = []
    barrier: set[int] = set()
    n = len(loop_points)
    for s in range(n):
        A, B = loop_points[s], loop_points[(s + 1) % n]
        forward = _lex_less(A, B) or (A[0] == B[0] and A[1] == B[1] and A[2] == B[2])
        S, E = (A, B) if forward else (B, A)
        dx, dy, dz = E[0] - S[0], E[1] - S[1], E[2] - S[2]
        chord = math.sqrt(dx * dx + dy * dy + dz * dz)
        steps = max(1, math.ceil(chord / spacing))
        raw = []
        for k in range(steps + 1):
            t = k / steps
            raw.append((E[0], E[1], E[2]) if k == steps else (S[0] + dx * t, S[1] + dy * t, S[2] + dz * t))
        ordered = raw[:steps] if forward else list(reversed(raw[1:]))
        for q in ordered:
            hit = closest(q)
            if hit is None:
                return {"error": "loop sample found no surface"}
            face = hit["triangle"] // 9
            barrier.add(face)
            pt = tuple(hit["point"])
            samples.append({"point": pt, "key": point_key(pt), "face": face, "bary": _barycentric(P, F, face, pt)})
    seed_hit = closest(seed)
    if seed_hit is None:
        return {"error": "seed found no surface"}
    seed_face = seed_hit["triangle"] // 9
    if seed_face in barrier:
        return {"error": "the loop passes through the seed face"}

    edge_faces = edge_face_map(F)
    flooded = {seed_face}
    queue = [seed_face]
    q = 0
    while q < len(queue):
        f = queue[q]
        q += 1
        for k in range(3):
            i, j = F[f * 3 + k], F[f * 3 + (k + 1) % 3]
            key = (i if i < j else j) * 16777216 + (j if i < j else i)
            for g in edge_faces[key]:
                if g == f or g in flooded or g in barrier:
                    continue
                flooded.add(g)
                queue.append(g)
    faces = [f for f in range(nf) if f in flooded or f in barrier]
    reach = 0.0
    for f in flooded:
        cx = (P[F[f * 3] * 3] + P[F[f * 3 + 1] * 3] + P[F[f * 3 + 2] * 3]) / 3 - seed[0]
        cy = (P[F[f * 3] * 3 + 1] + P[F[f * 3 + 1] * 3 + 1] + P[F[f * 3 + 2] * 3 + 1]) / 3 - seed[1]
        cz = (P[F[f * 3] * 3 + 2] + P[F[f * 3 + 1] * 3 + 2] + P[F[f * 3 + 2] * 3 + 2]) / 3 - seed[2]
        d = math.sqrt(cx * cx + cy * cy + cz * cz)
        if d > reach:
            reach = d
    return {"faces": faces, "samples": samples, "flooded": len(flooded), "barrier": len(barrier), "flood_reach_m": reach}


def loop_centroid_seed(closest, loop_points):
    x = y = z = 0.0
    for p in loop_points:
        x += p[0]
        y += p[1]
        z += p[2]
    n = len(loop_points)
    hit = closest((x / n, y / n, z / n))
    return tuple(hit["point"]) if hit else None


def split_loop_by_seam(loop_points, seam_points, snap_m: float = 0.015) -> dict:
    """Cut a closed outline into two panel loops along a seam whose ends lie on it.
    Port of splitLoopBySeam in flatten_core.mjs."""
    n = len(loop_points)
    if n < 3:
        return {"error": "outline needs at least 3 points"}
    if not seam_points or len(seam_points) < 2:
        return {"error": "seam needs at least 2 points"}

    def project(q):
        best = None
        for i in range(n):
            A, B = loop_points[i], loop_points[(i + 1) % n]
            dx, dy, dz = B[0] - A[0], B[1] - A[1], B[2] - A[2]
            ll = dx * dx + dy * dy + dz * dz
            t = ((q[0] - A[0]) * dx + (q[1] - A[1]) * dy + (q[2] - A[2]) * dz) / ll if ll > 0 else 0.0
            if t < 0:
                t = 0.0
            if t > 1:
                t = 1.0
            px = A[0] if t == 0 else B[0] if t == 1 else A[0] + dx * t
            py = A[1] if t == 0 else B[1] if t == 1 else A[1] + dy * t
            pz = A[2] if t == 0 else B[2] if t == 1 else A[2] + dz * t
            ex, ey, ez = q[0] - px, q[1] - py, q[2] - pz
            d = math.sqrt(ex * ex + ey * ey + ez * ez)
            if best is None or d < best["d"]:
                best = {"d": d, "i": i, "t": t, "point": (px, py, pz)}
        if best["t"] == 1:
            best["i"] = (best["i"] + 1) % n
            best["t"] = 0.0
        return best

    p1, p2 = project(seam_points[0]), project(seam_points[-1])
    if p1["d"] > snap_m or p2["d"] > snap_m:
        return {"error": f"seam end is {max(p1['d'], p2['d']) * 1000:.1f}mm from the outline (limit {snap_m * 1000}mm)"}

    def pos(p):
        return p["i"] + p["t"]

    if abs(pos(p1) - pos(p2)) < 1e-12:
        return {"error": "both seam ends land on the same outline point"}

    def forward(x, start):
        return ((x - start) % n + n) % n

    def between(a, b):
        out = []
        span = forward(pos(b), pos(a))
        for k in range(1, n):
            idx = (math.floor(pos(a)) + k) % n
            d = forward(idx, pos(a))
            if d >= span - 1e-12:
                break
            if d > 1e-12:
                out.append(loop_points[idx])
        return out

    interior = list(seam_points[1:-1])
    loop_a = [p1["point"]] + between(p1, p2) + [p2["point"]] + list(reversed(interior))
    loop_b = [p2["point"]] + between(p2, p1) + [p1["point"]] + interior
    return {"loops": [loop_a, loop_b], "split_points": [p1["point"], p2["point"]], "end_gaps_m": [p1["d"], p2["d"]]}


def loop_chords(samples, sub) -> dict:
    """Consecutive loop samples as barycentric distance constraints on a piece."""
    local_face = {g: i for i, g in enumerate(sub["face_ids"])}
    chords = []
    n = len(samples)
    for s in range(n):
        a, b = samples[s], samples[(s + 1) % n]
        fa, fb = local_face.get(a["face"]), local_face.get(b["face"])
        if fa is None or fb is None:
            return {"error": f"loop sample {s if fa is None else (s + 1) % n} lies outside the patch"}
        dx = b["point"][0] - a["point"][0]
        dy = b["point"][1] - a["point"][1]
        dz = b["point"][2] - a["point"][2]
        chords.append({"fa": fa, "ba": a["bary"], "fb": fb, "bb": b["bary"],
                       "rest": math.sqrt(dx * dx + dy * dy + dz * dz),
                       "pair": f"{a['key']}|{b['key']}" if a["key"] < b["key"] else f"{b['key']}|{a['key']}"})
    return {"chords": chords}


# --------------------------------------------------------------------- flatten

File: scripts/flatten_patch.mjs
/**
 * From a drawn loop to a patch: sampling the loop onto the mesh, flood-filling
 * its inside, turning the loop into chord constraints, cutting an outline into
 * two panels along a seam. The pen tool's input ends up here; the solver
 * (scripts/flatten_solver.mjs) never sees a loop, only chords. Port:
 * scripts/flatten_patch.py.
 */

import { edgeFaceMap } from './flatten_mesh.mjs';

export const DEFAULT_LOOP_SPACING = 0.002;   // barrier samples 2mm apart, < any edge

// ------------------------------------------------------------------ mesh build

function barycentric(P, F, face, q) {
  const i = F[face * 3] * 3, j = F[face * 3 + 1] * 3, k = F[face * 3 + 2] * 3;
  const v0x = P[j] - P[i], v0y = P[j + 1] - P[i + 1], v0z = P[j + 2] - P[i + 2];
  const v1x = P[k] - P[i], v1y = P[k + 1] - P[i + 1], v1z = P[k + 2] - P[i + 2];
  const v2x = q[0] - P[i], v2y = q[1] - P[i + 1], v2z = q[2] - P[i + 2];
  const d00 = v0x * v0x + v0y * v0y + v0z * v0z, d01 = v0x * v1x + v0y * v1y + v0z * v1z;
  const d11 = v1x * v1x + v1y * v1y + v1z * v1z, d20 = v2x * v0x + v2y * v0y + v2z * v0z;
  const d21 = v2x * v1x + v2y * v1y + v2z * v1z;
  const den = d00 * d11 - d01 * d01;
  const b1 = (d11 * d20 - d01 * d21) / den, b2 = (d00 * d21 - d01 * d20) / den;
  return [1 - b1 - b2, b1, b2];
}

/** A point's identity for matching the same loop sample across pieces: the
 *  same 3D coordinates produce the same key, so a seam two pieces share is
 *  recognised without any tolerance. */
export function pointKey(p) {
  return `${Math.floor(p[0] * 1e9 + 0.5)},${Math.floor(p[1] * 1e9 + 0.5)},${Math.floor(p[2] * 1e9 + 0.5)}`;
}

function lexLess(A, B) {
  if (A[0] !== B[0]) return A[0] < B[0];
  if (A[1] !== B[1]) return A[1] < B[1];
  return A[2] < B[2];
}

/**
 * Turn a closed loop drawn on the skin into a patch of faces.
 *
 * The loop is resampled at `spacing` (below any edge length, so consecutive
 * samples cannot skip a face), each sample snapped to the mesh, and every face a
 * sample lands on becomes a barrier. A flood fill over edge-adjacent faces from
 * the seed's face stops at the barrier; the patch is the flood plus the barrier,
 * so the loop lies entirely inside it and can be held to length by chord
 * constraints on the faces it crosses (see `loopChords`).
 *
 * Each segment is resampled in one canonical direction regardless of which way
 * the loop traverses it, so two pieces whose loops share a run of the same pen
 * line produce bit-identical samples along it — that is how a shared seam is
 * recognised.
 *
 * `closest` is a function p -> {point, normal, triangle} over the SAME soup the
 * mesh was welded from (scripts/surface_path.mjs closestOnMesh with its grid).
 * Declared limit: the ring of barrier faces is scaffolding that overshoots the
 * loop by up to one triangle; the piece's outline is the loop's image
 * (`mapLoopToFlat`), never the mesh boundary.
 */
export function extractPatch(mesh, closest, loopPoints, seed, spacing = DEFAULT_LOOP_SPACING) {
  const P = mesh.positions, F = mesh.faces;
  const nf = F.length / 3;
  const samples = [];
  const barrier = new Set();
  const n = loopPoints.length;
  for (let s = 0; s < n; s++) {
    const A = loopPoints[s], B = loopPoints[(s + 1) % n];
    const forward = lexLess(A, B) || (A[0] === B[0] && A[1] === B[1] && A[2] === B[2]);
    const S = forward ? A : B, E = forward ? B : A;
    const dx = E[0] - S[0], dy = E[1] - S[1], dz = E[2] - S[2];
    const chord = Math.sqrt(dx * dx + dy * dy + dz * dz);
    const steps = Math.max(1, Math.ceil(chord / spacing));
    const raw = [];
    for (let k = 0; k <= steps; k++) {
      const t = k / steps;
      raw.push(k === steps ? [E[0], E[1], E[2]] : [S[0] + dx * t, S[1] + dy * t, S[2] + dz * t]);
    }
    // the traversal keeps its start point and drops its end point
    const ordered = forward ? raw.slice(0, steps) : raw.slice(1).reverse();
    for (const q of ordered) {
      const hit = closest(q);
      if (!hit) return { error: 'loop sample found no surface' };
      const face = hit.triangle / 9;
      barrier.add(face);
      samples.push({ point: hit.point, key: pointKey(hit.point), face, bary: barycentric(P, F, face, hit.point) });
    }
  }
  const seedHit = closest(seed);
  if (!seedHit) return { error: 'seed found no surface' };
  const seedFace = seedHit.triangle / 9;
  if (barrier.has(seedFace)) return { error: 'the loop passes through the seed face' };

  const edgeFaces = edgeFaceMap(F);
  const flooded = new Set([seedFace]);
  const queue = [seedFace];
  for (let q = 0; q < queue.length; q++) {
    const f = queue[q];
    for (let k = 0; k < 3; k++) {
      const i = F[f * 3 + k], j = F[f * 3 + ((k + 1) % 3)];
      const key = (i < j ? i : j) * 16777216 + (i < j ? j : i);
      for (const g of edgeFaces.get(key)) {
        if (g === f || flooded.has(g) || barrier.has(g)) continue;
        flooded.add(g); queue.push(g);
      }
    }
  }
  const faces = [];
  for (let f = 0; f < nf; f++) if (flooded.has(f) || barrier.has(f)) faces.push(f);
  // how far the flood reached from the seed: a leak into the rest of the body
  // shows up here as a distance far beyond the loop
  let reach = 0;
  for (const f of flooded) {
    const cx = (P[F[f * 3] * 3] + P[F[f * 3 + 1] * 3] + P[F[f * 3 + 2] * 3]) / 3 - seed[0];
    const cy = (P[F[f * 3] * 3 + 1] + P[F[f * 3 + 1] * 3 + 1] + P[F[f * 3 + 2] * 3 + 1]) / 3 - seed[1];
    const cz = (P[F[f * 3] * 3 + 2] + P[F[f * 3 + 1] * 3 + 2] + P[F[f * 3 + 2] * 3 + 2]) / 3 - seed[2];
    const d = Math.sqrt(cx * cx + cy * cy + cz * cz);
    if (d > reach) reach = d;
  }
  return { faces, samples, flooded: flooded.size, barrier: barrier.size, flood_reach_m: reach };
}

/** A seed for `extractPatch`: the loop's centroid snapped to the surface. For
 *  the convex-ish loops a pattern piece is, that lands inside; if it does not,
 *  the flood fill's reach says so and the caller must refuse. */
export function loopCentroidSeed(closest, loopPoints) {
  let x = 0, y = 0, z = 0;
  for (const p of loopPoints) { x += p[0]; y += p[1]; z += p[2]; }
  const n = loopPoints.length;
  const hit = closest([x / n, y / n, z / n]);
  return hit ? hit.point : null;
}

/**
 * Cut a closed outline into two panel loops along a seam polyline whose ends lie
 * on (or within `snap_m` of) the outline. Each seam end is projected onto the
 * nearest outline segment and that projected point becomes a vertex of both
 * panels; the seam's own end points are dropped in its favour, so the barrier
 * the two loops lay down is closed exactly and the run they share consists of
 * the very same points — which is how `relaxPieces` recognises a shared seam.
 *
 * Returns { loops: [a, b], split_points, end_gaps_m } or { error }.
 */
export function splitLoopBySeam(loopPoints, seamPoints, snap_m = 0.015) {
  const n = loopPoints.length;
  if (n < 3) return { error: 'outline needs at least 3 points' };
  if (!seamPoints || seamPoints.length < 2) return { error: 'seam needs at least 2 points' };
  const project = (q) => {
    let best = null;
    for (let i = 0; i < n; i++) {
      const A = loopPoints[i], B = loopPoints[(i + 1) % n];
      const dx = B[0] - A[0], dy = B[1] - A[1], dz = B[2] - A[2];
      const ll = dx * dx + dy * dy + dz * dz;
      let t = ll > 0 ? ((q[0] - A[0]) * dx + (q[1] - A[1]) * dy + (q[2] - A[2]) * dz) / ll : 0;
      if (t < 0) t = 0; if (t > 1) t = 1;
      const px = t === 0 ? A[0] : t === 1 ? B[0] : A[0] + dx * t;
      const py = t === 0 ? A[1] : t === 1 ? B[1] : A[1] + dy * t;
      const pz = t === 0 ? A[2] : t === 1 ? B[2] : A[2] + dz * t;
      const ex = q[0] - px, ey = q[1] - py, ez = q[2] - pz;
      const d = Math.sqrt(ex * ex + ey * ey + ez * ez);
      if (!best || d < best.d) best = { d, i, t, point: [px, py, pz] };
    }
    if (best.t === 1) { best.i = (best.i + 1) % n; best.t = 0; }   // one name for a shared corner
    return best;
  };
  const p1 = project(seamPoints[0]), p2 = project(seamPoints[seamPoints.length - 1]);
  if (p1.d > snap_m || p2.d > snap_m) return { error: `seam end is ${(Math.max(p1.d, p2.d) * 1000).toFixed(1)}mm from the outline (limit ${snap_m * 1000}mm)` };
  const pos = (p) => p.i + p.t;
  if (Math.abs(pos(p1) - pos(p2)) < 1e-12) return { error: 'both seam ends land on the same outline point' };
  // outline vertices strictly between two split positions, walking forward
  const forward = (x, from) => (((x - from) % n) + n) % n;
  const between = (from, to) => {
    const out = [];
    const span = forward(pos(to), pos(from));
    for (let k = 1; k < n; k++) {
      const idx = (Math.floor(pos(from)) + k) % n;
      const d = forward(idx, pos(from));
      if (d >= span - 1e-12) break;
      if (d > 1e-12) out.push(loopPoints[idx]);
    }
    return out;
  };
  const interior = seamPoints.slice(1, -1);
  const arcAB = between(p1, p2), arcBA = between(p2, p1);
  const loopA = [p1.point, ...arcAB, p2.point, ...interior.slice().reverse()];
  const loopB = [p2.point, ...arcBA, p1.point, ...interior];
  return { loops: [loopA, loopB], split_points: [p1.point, p2.point], end_gaps_m: [p1.d, p2.d] };
}

/**
 * The chords of a drawn loop as constraints on a piece: consecutive samples,
 * each a barycentric point in a local face, held to their 3D chord length.
 * `pair` names the chord by its two sample keys, orientation-free, so the same
 * chord in another piece is recognised as shared.
 */
export function loopChords(samples, sub) {
  const localFace = new Map();
  sub.faceIds.forEach((g, i) => localFace.set(g, i));
  const chords = [];
  const n = samples.length;
  for (let s = 0; s < n; s++) {
    const a = samples[s], b = samples[(s + 1) % n];
    const fa = localFace.get(a.face), fb = localFace.get(b.face);
    if (fa === undefined || fb === undefined) return { error: `loop sample ${fa === undefined ? s : (s + 1) % n} lies outside the patch` };
    const dx = b.point[0] - a.point[0], dy = b.point[1] - a.point[1], dz = b.point[2] - a.point[2];
    chords.push({
      fa, ba: a.bary, fb, bb: b.bary,
      rest: Math.sqrt(dx * dx + dy * dy + dz * dz),
      pair: a.key < b.key ? `${a.key}|${b.key}` : `${b.key}|${a.key}`,
    });
  }
  return { chords };
}

// --------------------------------------------------------------------- flatten
